wtforms/fields/core.py

```diff
@@ -472,6 +472,12 @@ class SelectField(SelectFieldBase):
     def __init__(self, label=None, validators=None, coerce=text_type, choices=None, validate_choice=True, **kwargs):
         super(SelectField, self).__init__(label, validators, **kwargs)
         self.coerce = coerce
+        if callable(choices):
+            choices = choices()
+        if isinstance(choices, dict):
+            self.groups = choices
+        else:
+            self.groups = {}
         self.choices = list(choices) if choices is not None else None
```

Comment:
`self.choices` will not be set to the expected value when a dict is passed in. This still needs to remain consistent (the optgroup doesn't affect the accepted values, only how they are presented).